src/control/ros/RosTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,290 @@
+      return "goal tolerance violated";
+
+    default:
+      return "unknown";
+  }
+}
+
+} // namespace
+
+//=============================================================================
+RosTrajectoryExecutor::RosTrajectoryExecutor(
+      ::dart::dynamics::MetaSkeletonPtr skeleton, 
+      ::ros::NodeHandle node,
+      const std::string& serverName,
+      double timestep,
+      double goalTimeTolerance,
+      std::chrono::milliseconds connectionTimeout,
+      std::chrono::milliseconds connectionPollingRate)
+  : mNode{std::move(node)}
+  , mCallbackQueue{}
+  , mClient{mNode, serverName, &mCallbackQueue}
+  , mSkeleton{std::move(skeleton)}
+  , mTimestep{timestep}
+  , mGoalTimeTolerance{goalTimeTolerance}
+  , mConnectionTimeout{connectionTimeout}
+  , mConnectionPollingRate{connectionPollingRate}
+  , mInProgress{false}
+{
+  if (!mSkeleton)
+    throw std::invalid_argument("Skeleton is null.");
+
+  if (mTimestep <= 0)
+    throw std::invalid_argument("Timestep must be positive.");
+
+  if (mGoalTimeTolerance <= 0)
+    throw std::invalid_argument("Goal time tolerance must be positive.");
+}
+
+//=============================================================================
+RosTrajectoryExecutor::~RosTrajectoryExecutor()
+{
+  // Do nothing.
+  // TODO: Should we wait for the current trajectory to finish executing?
+}
+
+//=============================================================================
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr _traj)
+{
+  static const ::ros::Time invalidTime;
+  return execute(_traj, invalidTime);
+}
+
+//=============================================================================
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr _traj, const ::ros::Time& _startTime)
+{
+  using statespace::dart::MetaSkeletonStateSpace;
+
+  if (!_traj)
+    throw std::invalid_argument("Trajectory is null.");
+
+  const auto space = std::dynamic_pointer_cast<
+    MetaSkeletonStateSpace>(_traj->getStateSpace());
+  if (!space)
+  {
+    throw std::invalid_argument(
+      "Trajectory is not in a MetaSkeletonStateSpace.");
+  }
+
+  const auto trajectorySkeleton = space->getMetaSkeleton();
+  for (const auto dof : mSkeleton->getDofs())
+  {
+    const auto idof = trajectorySkeleton->getIndexOf(dof, false);
+    if (idof == dart::dynamics::INVALID_INDEX)
+    {
+      std::stringstream msg;
+      msg << "Trajectory is missing DegreeOfFreedom '" << dof->getName()
+          << "' for Skeleton '" << dof->getSkeleton()->getName() << "'.";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+  // Setup the goal properties.
+  // TODO: Also set the goal properties.
+  control_msgs::FollowJointTrajectoryGoal goal;
+  goal.trajectory.header.stamp = _startTime;
+  goal.goal_time_tolerance = ::ros::Duration(mGoalTimeTolerance);
+
+  // Convert the Aikido trajectory into a ROS JointTrajectory.
+  // TODO: Move this into a helper function.
```

Comment:
We should move this into a helper function. Otherwise, it will be extremely difficult to unit test this function.